### PR TITLE
Change default apple_platform_type to macOS

### PR DESCRIPTION
**Pull request by keith**
When using rules for Apple platforms that aren't platform specific, such
as `swift_binary`, the default of iOS doesn't make sense, instead we
should default this to the current platform, which currently can only be
macOS.

More details about this issue can be found here https://github.com/bazelbuild/rules_swift/issues/51

### CODE REVIEW

**Comment by allevato** [comment]
This change would have very broad ranging implications within Google. I believe there is a _significant_ amount of cleanup that would need to be done before we could flip this switch, which is why we haven't done it yet.

**Comment by keith** [comment]
What other rules are affected by this?

**Comment by allevato** [comment]
All the `objc_*` rules would be affected by this.

The reason for the iOS default is entirely historical (the first users of Apple support in Bazel were building iOS apps, and today it's still the case that that's the majority of our users), so it "made sense" at the time (in fact, iOS was the only supported target for a while).

It's not so much the rules themselves that matter here, but all the implicit assumptions elsewhere that assume that a flagless build will target iOS that would break.

**Comment by lberki** [comment]
I only touched `AppleCommandLineOptions` only due to my interest in C++ rules and because I did a cleanup out of general desire for cleanliness.

I think it would be worthwhile to eventually unify the configuration distinguisher and the platform type, but that would require adding the "unspecified Apple" platform type. But since I don't actually know much about Apple rules, much less about how this change would affect Google-internal use cases, I'll defer to Sergio.

**Comment by keith** [comment]
Does this mean google internally is using this undocumented flag as well? It seems like we should either add a way for rules to specify this, or publicize this flag

**Comment by ulfjack** [comment]
Yes, the flag is definitely used. It seems to be set on an ios config (in a bazelrc); if everyone is using that, changing the default would be safe.

**Comment by keith** [comment]
I assume there would be a problem there too where if you had a target that you built for macOS for tools, it would need to have that overridden (but maybe that's already happening so it would be ok)

**Comment by keith** [comment]
Any thoughts on how we can get to the point where you can build some targets like `swift_binary` without having to specify this flag? Maybe an alternative is to allow setting the platform type (or just defaulting it) on those rules?

**Comment by sergiocampama** [comment]
One thing that the bazel team could do is to copybara this change so that bazel by default is MACOS, while internally it's still IOS. Another thing would be to make it MACOS by default and add IOS as the default in bazelrc files.

**Comment by keith** [comment]
@aragos do you have any thoughts on the possible solutions here?

**Comment by keith** [comment]
I've updated the tests that were relying on the iOS default to pass `--apple_platform_type` explicitly.

File: src/test/java/com/google/devtools/build/lib/rules/objc/AppleToolchainSelectionTest.java [hunk 1]
```diff
@@ -41,12 +41,12 @@ public void testToolchainSelectionDefault() throws Exception {
 
     assertThat(cppConfig.getRuleProvidingCcToolchainProvider().toString())
         .isEqualTo("//tools/osx/crosstool:crosstool");
-    assertThat(cppConfig.getTransformedCpuFromOptions()).isEqualTo("ios_x86_64");
+    assertThat(cppConfig.getTransformedCpuFromOptions()).isEqualTo("darwin_x86_64");
```
Comment (keith) [inline src/test/java/com/google/devtools/build/lib/rules/objc/AppleToolchainSelectionTest.java]: This is the only test I intentionally changed the behavior of, I believe that we want to now test the default is macOS (and therefore is this default arch)

**Comment by aragos** [comment]
Looks good to me in principle, I can take care of explicitly setting the Google-internal default so that this change doesn't cause any behavioral differences there. 

We should also test (and fix) any other downstream projects that may be affected, this can be done following this guide: https://github.com/bazelbuild/continuous-integration/blob/master/docs/downstream-testing.md

Finally there still appear to be some failing tests in Bazel's buildkite, please fix those as well.

Once the two buildkite verifications are done I'll import the PR.

**Comment by keith** [comment]
Sounds good. Quick question @aragos should I follow those word for word and create a new branch with this same change, or can we trigger that build on this branch when we're ready?

**Comment by aragos** [comment]
@keith I believe we'll need a fresh branch with this same change. Feel free to create it and run the downstream projects if you have the necessary permissions, otherwise let me know and I can help.

**Comment by keith** [comment]
Sounds good. I've emailed bazel_dev asking for the right permissions. If I don't get them by the time I'm ready I'll ask, thanks!

**Comment by keith** [comment]
Oh actually, the recommendation is to create that branch on the bazel repo itself, not my fork, which isn't something I can do without having push access to this repo. I've pushed a branch with the naming suggested in that doc to my fork https://github.com/bazelbuild/bazel/compare/master...keith:keith-test-feature-platform-default @aragos is this something you can mirror to the bazel repo so we can run the tests (we might want to wait until all other tests are green here first though)

**Comment by keith** [comment]
Ok, it passed, so @aragos I guess we're good to start the others

**Comment by aragos** [comment]
Great! I imported and pushed your changes to https://github.com/bazelbuild/bazel/tree/aragos-test-macos-default-platform and started a buildkite run with that at https://buildkite.com/bazel/bazel-at-head-plus-downstream/builds/735. This may take a while, let's see how it goes.

**Comment by keith** [comment]
Thanks! I don't have access yet but hopefully I'll get it soon. 


--
Keith Smiley

**Comment by aragos** [comment]
Downstream tests have passed. I'm currently working on preserving the old default internally then I'll merge the PR.

**Comment by keith** [comment]
🎉 thanks a ton everyone!

**Pull request closed without merging**